File: c/サンプル/02_成績管理システム.c
/*
 * ============================================================
 *  成績管理システム
 * ============================================================
 *
 *  【学べる内容】
 *    - 配列の宣言と操作
 *    - for / while による繰り返し処理
 *    - if-else による条件分岐
 *    - 関数の定義と呼び出し
 *    - 平均・最大・最小の求め方
 *
 *  【実行方法】
 *    gcc -o output 02_成績管理システム.c && ./output
 *
 * ============================================================
 */

#include <stdio.h>
#include <string.h>

#define MAX_STUDENTS 50   /* 最大生徒数 */
#define MAX_NAME     32   /* 名前の最大文字数 */

/* --- 関数プロトタイプ --- */
double calc_average(const int scores[], int count);
int    find_max(const int scores[], int count);
int    find_min(const int scores[], int count);
char   get_grade(int score);
void   print_report(const char names[][MAX_NAME], const int scores[], int count);

/* ===== メイン関数 ===== */
int main(void)
{
    char names[MAX_STUDENTS][MAX_NAME];  /* 生徒の名前 */
    int  scores[MAX_STUDENTS];           /* 各生徒の点数 */
    int  count = 0;                      /* 登録人数 */
    int  choice;

    printf("==========================================\n");
    printf("        成績管理システム\n");
    printf("==========================================\n");

    /* --- メニューループ --- */
    while (1) {
        printf("\n--- メニュー ---\n");
        printf("  1. 生徒の成績を登録\n");
        printf("  2. 成績一覧を表示\n");
        printf("  3. 統計情報を表示\n");
        printf("  4. 終了\n");
        printf("選択してください (1-4): ");

        if (scanf("%d", &choice) != 1) {
            /* 数値以外が入力された場合 */
            while (getchar() != '\n');
            printf("※ 1〜4 の数字を入力してください。\n");
            continue;
        }
        while (getchar() != '\n');

        switch (choice) {
        /* ---------- 1. 成績登録 ---------- */
        case 1:
            if (count >= MAX_STUDENTS) {
                printf("※ これ以上登録できません（最大 %d 人）。\n", MAX_STUDENTS);
                break;
            }
            printf("名前を入力してください: ");
            scanf("%31[^\n]", names[count]);
            while (getchar() != '\n');

            printf("点数を入力してください (0-100): ");
            scanf("%d", &scores[count]);
            while (getchar() != '\n');

            if (scores[count] < 0 || scores[count] > 100) {
                printf("※ 点数は 0〜100 の範囲で入力してください。登録をキャンセルしました。\n");
                break;
            }

            printf(">> %s さん（%d 点）を登録しました。\n",
                   names[count], scores[count]);
            count++;
            break;

        /* ---------- 2. 一覧表示 ---------- */
        case 2:
            if (count == 0) {
                printf("※ まだ生徒が登録されていません。\n");
                break;
            }
            print_report(names, scores, count);
            break;

        /* ---------- 3. 統計情報 ---------- */
        case 3:
            if (count == 0) {
                printf("※ まだ生徒が登録されていません。\n");
                break;
            }
            {
                double avg = calc_average(scores, count);
                int    max = find_max(scores, count);
                int    min = find_min(scores, count);

                printf("\n========== 統計情報 ==========\n");
                printf("  登録人数 : %d 人\n", count);
                printf("  平均点   : %.1f 点\n", avg);
                printf("  最高点   : %d 点\n", max);
                printf("  最低点   : %d 点\n", min);

                /* 成績分布 */
                int grade_count[5] = {0}; /* A, B, C, D, F */
                for (int i = 0; i < count; i++) {
                    switch (get_grade(scores[i])) {
                    case 'A': grade_count[0]++; break;
                    case 'B': grade_count[1]++; break;
                    case 'C': grade_count[2]++; break;
                    case 'D': grade_count[3]++; break;
                    case 'F': grade_count[4]++; break;
                    }
                }
                printf("\n  --- 成績分布 ---\n");
                printf("  A (90-100) : %d 人\n", grade_count[0]);
                printf("  B (80-89)  : %d 人\n", grade_count[1]);
                printf("  C (70-79)  : %d 人\n", grade_count[2]);
                printf("  D (60-69)  : %d 人\n", grade_count[3]);
                printf("  F ( 0-59)  : %d 人\n", grade_count[4]);
                printf("==============================\n");
            }
            break;

        /* ---------- 4. 終了 ---------- */
        case 4:
            printf("成績管理システムを終了します。\n");
            return 0;

        default:
            printf("※ 1〜4 の数字を入力してください。\n");
            break;
        }
    }

    return 0;
}

/* ===== 平均点を計算する関数 ===== */
double calc_average(const int scores[], int count)
{
    int sum = 0;
    for (int i = 0; i < count; i++) {
        sum += scores[i];
    }
    return (double)sum / count;   /* int → double へキャスト */
}

/* ===== 最高点を求める関数 ===== */
int find_max(const int scores[], int count)
{
    int max = scores[0];
    for (int i = 1; i < count; i++) {
        if (scores[i] > max) {
            max = scores[i];
        }
    }
    return max;
}

/* ===== 最低点を求める関数 ===== */
int find_min(const int scores[], int count)
{
    int min = scores[0];
    for (int i = 1; i < count; i++) {
        if (scores[i] < min) {
            min = scores[i];
        }
    }
    return min;
}

/* ===== 点数から成績（A〜F）を返す関数 ===== */
char get_grade(int score)
{
    if (score >= 90) return 'A';
    if (score >= 80) return 'B';
    if (score >= 70) return 'C';
    if (score >= 60) return 'D';
    return 'F';
}

/* ===== 成績レポートを表示する関数 ===== */
void print_report(const char names[][MAX_NAME], const int scores[], int count)
{
    printf("\n+------+------------------+------+------+\n");
    printf("| No.  | 名前             | 点数 | 評価 |\n");
    printf("+------+------------------+------+------+\n");

    for (int i = 0; i < count; i++) {
        char grade = get_grade(scores[i]);
        printf("| %4d | %-16s | %4d |   %c  |\n",
               i + 1, names[i], scores[i], grade);
    }

    printf("+------+------------------+------+------+\n");
}

File: java/サンプル/02_成績管理システム.java
/**
 * ============================================
 *   成績管理システム
 * ============================================
 *
 * 【学べる内容】
 *   - 配列（array）と ArrayList の使い方
 *   - メソッド（method）の定義・引数・戻り値
 *   - for ループ・拡張for文（for-each）
 *   - 条件分岐（if-else, switch）
 *   - 文字列のフォーマット出力
 *   - メニュー方式のプログラム設計
 *   - Collections クラスの活用
 *
 * 【実行方法】
 *   javac 02_成績管理システム.java
 *   java GradeManager
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class GradeManager {

    // ========== フィールド（クラス変数） ==========
    // 学生の名前を格納するリスト
    static ArrayList<String> studentNames = new ArrayList<>();
    // 学生の点数を格納するリスト（名前と同じインデックスで対応）
    static ArrayList<Integer> studentScores = new ArrayList<>();

    // ========== メインメソッド ==========
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("╔══════════════════════════════════════╗");
        System.out.println("║       成績管理システム v1.0          ║");
        System.out.println("╚══════════════════════════════════════╝");
        System.out.println();

        // サンプルデータを登録（デモ用）
        addSampleData();

        // メニューループ: ユーザーが「6」を入力するまで繰り返す
        boolean running = true;
        while (running) {
            displayMenu();
            System.out.print("選択してください (1-6): ");
            String choice = scanner.nextLine().trim();

            // switch文で選択に応じた処理を実行
            switch (choice) {
                case "1":
                    addStudent(scanner);
                    break;
                case "2":
                    viewAllStudents();
                    break;
                case "3":
                    showStatistics();
                    break;
                case "4":
                    findTopStudent();
                    break;
                case "5":
                    searchStudent(scanner);
                    break;
                case "6":
                    running = false;
                    System.out.println("\nシステムを終了します。お疲れさまでした！");
                    break;
                default:
                    System.out.println("\n※ 1〜6の数字を入力してください。");
            }
        }

        scanner.close();
    }

    /**
     * メニューを表示するメソッド
     */
    static void displayMenu() {
        System.out.println("\n┌──────────── メニュー ────────────┐");
        System.out.println("│  1. 学生を追加する               │");
        System.out.println("│  2. 全学生の成績を表示           │");
        System.out.println("│  3. 統計情報を表示               │");
        System.out.println("│  4. 最高得点者を表示             │");
        System.out.println("│  5. 学生を検索する               │");
        System.out.println("│  6. 終了                         │");
        System.out.println("└──────────────────────────────────┘");
    }

    /**
     * サンプルデータを追加するメソッド
     */
    static void addSampleData() {
        // 配列を使ってサンプルデータを定義
        String[] names = {"田中太郎", "鈴木花子", "佐藤次郎", "山田美咲", "高橋一郎"};
        int[] scores = {85, 92, 78, 95, 88};

        // 拡張for文（for-each）ではなく、インデックスが必要なので通常のfor文を使用
        for (int i = 0; i < names.length; i++) {
            studentNames.add(names[i]);
            studentScores.add(scores[i]);
        }

        System.out.printf("サンプルデータ %d 件を登録しました。%n", names.length);
    }

    /**
     * 学生を追加するメソッド
     *
     * @param scanner 入力用Scannerオブジェクト
     */
    static void addStudent(Scanner scanner) {
        System.out.println("\n--- 学生の追加 ---");

        System.out.print("学生の名前: ");
        String name = scanner.nextLine().trim();

        // 名前が空でないかチェック
        if (name.isEmpty()) {
            System.out.println("※ 名前を入力してください。");
            return; // メソッドを早期リターン
        }

        System.out.print("点数 (0-100): ");
        int score;
        try {
            score = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException e) {
            // 数値でない入力への対処
            System.out.println("※ 数値を入力してください。");
            return;
        }

        // 点数の範囲チェック
        if (score < 0 || score > 100) {
            System.out.println("※ 0〜100の範囲で入力してください。");
            return;
        }

        // ArrayListに追加
        studentNames.add(name);
        studentScores.add(score);

        System.out.printf("%s さん（%d 点）を登録しました。%n", name, score);
    }

    /**
     * 全学生の成績を一覧表示するメソッド
     */
    static void viewAllStudents() {
        System.out.println("\n--- 全学生の成績一覧 ---");

        // データがない場合のチェック
        if (studentNames.isEmpty()) {
            System.out.println("※ 登録されている学生がいません。");
            return;
        }

        // テーブルヘッダー
        System.out.println("┌────┬──────────┬──────┬──────┐");
        System.out.println("│ No │  名前    │ 点数 │ 評価 │");
        System.out.println("├────┼──────────┼──────┼──────┤");

        // 拡張for文は使えない（インデックスが必要なため）ので通常のforを使用
        for (int i = 0; i < studentNames.size(); i++) {
            String name = studentNames.get(i);
            int score = studentScores.get(i);
            String grade = getGrade(score);

            // printf でフォーマット出力
            System.out.printf("│ %2d │ %-8s │ %3d  │  %s  │%n",
                    i + 1, name, score, grade);
        }

        System.out.println("└────┴──────────┴──────┴──────┘");
        System.out.printf("登録人数: %d 名%n", studentNames.size());
    }

    /**
     * 統計情報を表示するメソッド
     */
    static void showStatistics() {
        System.out.println("\n--- 統計情報 ---");

        if (studentScores.isEmpty()) {
            System.out.println("※ データがありません。");
            return;
        }

        // 平均点を計算
        double average = calculateAverage();
        // 最高点と最低点を取得（Collectionsクラスの便利メソッド）
        int maxScore = Collections.max(studentScores);
        int minScore = Collections.min(studentScores);

        // 評価ごとの人数を数える
        int countA = 0, countB = 0, countC = 0, countD = 0, countF = 0;
        for (int score : studentScores) {  // 拡張for文（for-each）
            switch (getGrade(score)) {
                case "A": countA++; break;
                case "B": countB++; break;
                case "C": countC++; break;
                case "D": countD++; break;
                case "F": countF++; break;
            }
        }

        System.out.printf("  受験者数 : %d 名%n", studentScores.size());
        System.out.printf("  平均点   : %.1f 点%n", average);
        System.out.printf("  最高点   : %d 点%n", maxScore);
        System.out.printf("  最低点   : %d 点%n", minScore);
        System.out.printf("  点数幅   : %d 点%n", maxScore - minScore);
        System.out.println();
        System.out.println("  【評価別人数】");
        System.out.printf("    A (90-100): %d 名  ", countA);
        printBar(countA, studentScores.size());
        System.out.printf("    B (80-89) : %d 名  ", countB);
        printBar(countB, studentScores.size());
        System.out.printf("    C (70-79) : %d 名  ", countC);
        printBar(countC, studentScores.size());
        System.out.printf("    D (60-69) : %d 名  ", countD);
        printBar(countD, studentScores.size());
        System.out.printf("    F (0-59)  : %d 名  ", countF);
        printBar(countF, studentScores.size());
    }

    /**
     * 簡易棒グラフを表示する
     */
    static void printBar(int count, int total) {
        int barLength = (total > 0) ? (count * 20 / total) : 0;
        System.out.println("█".repeat(barLength));
    }

    /**
     * 最高得点者を表示するメソッド
     */
    static void findTopStudent() {
        System.out.println("\n--- 最高得点者 ---");

        if (studentScores.isEmpty()) {
            System.out.println("※ データがありません。");
            return;
        }

        int maxScore = Collections.max(studentScores);

        System.out.printf("最高点: %d 点%n", maxScore);
        System.out.println("該当者:");

        // 最高点の学生が複数いる場合を考慮
        for (int i = 0; i < studentScores.size(); i++) {
            if (studentScores.get(i) == maxScore) {
                System.out.printf("  ★ %s さん%n", studentNames.get(i));
            }
        }
    }

    /**
     * 名前で学生を検索するメソッド
     *
     * @param scanner 入力用Scannerオブジェクト
     */
    static void searchStudent(Scanner scanner) {
        System.out.println("\n--- 学生検索 ---");
        System.out.print("検索する名前（部分一致）: ");
        String keyword = scanner.nextLine().trim();

        if (keyword.isEmpty()) {
            System.out.println("※ キーワードを入力してください。");
            return;
        }

        boolean found = false;
        for (int i = 0; i < studentNames.size(); i++) {
            // String.contains() で部分一致検索
            if (studentNames.get(i).contains(keyword)) {
                if (!found) {
                    System.out.println("検索結果:");
                    found = true;
                }
                System.out.printf("  %s : %d 点 (%s)%n",
                        studentNames.get(i),
                        studentScores.get(i),
                        getGrade(studentScores.get(i)));
            }
        }

        if (!found) {
            System.out.printf("「%s」に一致する学生は見つかりませんでした。%n", keyword);
        }
    }

    // ========== ユーティリティメソッド ==========

    /**
     * 平均点を計算して返すメソッド
     *
     * @return 平均点（double型）
     */
    static double calculateAverage() {
        int sum = 0;
        // 拡張for文で全要素を合計
        for (int score : studentScores) {
            sum += score;
        }
        // int同士の割り算は小数が切り捨てられるため、doubleにキャスト
        return (double) sum / studentScores.size();
    }

    /**
     * 点数から評価（グレード）を返すメソッド
     *
     * @param score 点数（0-100）
     * @return 評価文字列
     */
    static String getGrade(int score) {
        if (score >= 90) return "A";
        else if (score >= 80) return "B";
        else if (score >= 70) return "C";
        else if (score >= 60) return "D";
        else return "F";
    }
}
